### ISSUE TITLE
Getting NaN loss when adding a split layer

### ISSUE THREAD

**Issue by krumo** [facebookresearch/Detectron#582]
### Expected results

I want to add an unsupervised learning part for Detectron. Thus, I plan to feed a labeled image and an unlabeled image in each batch to GPU and to compute bbox and cls loss only for labeled images. To verify I am on the right way, I just modify the given end to end faster R-CNNN network part by adding a split layer to split the cls_logits, bbox_pred and other related labels and weights. Then I set imgs_per_batch as 2 and train the network with labeled cityscapes training set. Ideally I could get a trained model on a subset of cityscapes training set.

### Actual results

However, after only one iteration, the loss for rpn_bbox and rpn_cls becomes NaN and the training stops. Following is the terminal output for loss.

> json_stats: {"accuracy_cls": 0.000000, "eta": "1 day, 22:21:25", "iter": 1, "loss": NaN, "loss_bbox": 0.183388, "loss_cls": 2.307324, "loss_rpn_bbox": NaN, "loss_rpn_cls": NaN, "lr": 0.001000, "mb_qsize": 64, "mem": 2593, "time": 2.384116}

What's more interesting, when I add a stopgradient layer to `rpn_cls_logits`, this error disappears.

### Detailed steps to reproduce

Modify the  `add_single_scale_rpn_losses` function in rpn_heads.py:

```
def add_single_scale_rpn_losses(model):
    """Add losses for a single scale RPN model (i.e., no FPN)."""
    # Spatially narrow the full-sized RPN label arrays to match the feature map
    # shape
    model.net.SpatialNarrowAs(
        ['rpn_labels_int32_wide', 'rpn_cls_logits'], 'rpn_labels_int32'
    )
    for key in ('targets', 'inside_weights', 'outside_weights'):
        model.net.SpatialNarrowAs(
            ['rpn_bbox_' + key + '_wide', 'rpn_bbox_pred'], 'rpn_bbox_' + key
        )
    model.net.Split('rpn_cls_logits', ['labeled_rpn_cls_logits','unlabeled_rpn_cls_logits'], axis=0)
    model.net.Split('rpn_bbox_pred', ['labeled_rpn_bbox_pred','unlabeled_rpn_bbox_pred'], axis=0)
    model.net.Split('rpn_labels_int32', ['labeled_rpn_labels_int32','unlabeled_rpn_labels_int32'], axis=0)
    model.net.Split('rpn_bbox_targets', ['labeled_rpn_bbox_targets','unlabeled_rpn_bbox_targets'], axis=0)
    model.net.Split('rpn_bbox_inside_weights', ['labeled_rpn_bbox_inside_weights','unlabeled_rpn_bbox_inside_weights'], axis=0)
    model.net.Split('rpn_bbox_outside_weights', ['labeled_rpn_bbox_outside_weights','unlabeled_rpn_bbox_outside_weights'], axis=0)

    # model.StopGradient('labeled_rpn_cls_logits', 'labeled_rpn_cls_logits')
    loss_rpn_cls = model.net.SigmoidCrossEntropyLoss(
        ['labeled_rpn_cls_logits', 'labeled_rpn_labels_int32'],
        'loss_rpn_cls',
        scale=model.GetLossScale()
    )
    loss_rpn_bbox = model.net.SmoothL1Loss(
        [
            'labeled_rpn_bbox_pred', 'labeled_rpn_bbox_targets', 'labeled_rpn_bbox_inside_weights',
            'labeled_rpn_bbox_outside_weights'
        ],
        'loss_rpn_bbox',
        beta=1. / 9.,
        scale=model.GetLossScale()
    )
    loss_gradients = blob_utils.get_loss_gradients(
        model, [loss_rpn_cls, loss_rpn_bbox]
    )
    model.AddLosses(['loss_rpn_cls', 'loss_rpn_bbox'])
    return loss_gradients
```
And modify the  `add_fast_rcnn_losses` function in fast_rcnn_heads.py:
```
def add_fast_rcnn_losses(model):
    """Add losses for RoI classification and bounding box regression."""
    model.net.Split('cls_score', ['labeled_cls_score','unlabeled_cls_score'], axis=0)
    model.net.Split('bbox_pred', ['labeled_bbox_pred','unlabeled_bbox_pred'], axis=0)
    model.net.Split('labels_int32', ['labeled_labels_int32','unlabeled_labels_int32'], axis=0)
    model.net.Split('bbox_targets', ['labeled_bbox_targets','unlabeled_bbox_targets'], axis=0)
    model.net.Split('bbox_inside_weights', ['labeled_bbox_inside_weights','unlabeled_bbox_inside_weights'], axis=0)
    model.net.Split('bbox_outside_weights', ['labeled_bbox_outside_weights','unlabeled_bbox_outside_weights'], axis=0)

    cls_prob, loss_cls = model.net.SoftmaxWithLoss(
        ['labeled_cls_score', 'labeled_labels_int32'], ['cls_prob', 'loss_cls'],
        scale=model.GetLossScale()
    )
    loss_bbox = model.net.SmoothL1Loss(
        [
            'labeled_bbox_pred', 'labeled_bbox_targets', 'labeled_bbox_inside_weights',
            'labeled_bbox_outside_weights'
        ],
        'loss_bbox',
        scale=model.GetLossScale()
    )
    loss_gradients = blob_utils.get_loss_gradients(model, [loss_cls, loss_bbox])
    model.Accuracy(['cls_prob', 'labeled_labels_int32'], 'accuracy_cls')
    model.AddLosses(['loss_cls', 'loss_bbox'])
    model.AddMetrics('accuracy_cls')
    return loss_gradients
```
And then run train_net.py with an end to end RPN based Faster R-CNN configuration setting IMS_PER_BATCH as 2.
### System information

* Operating system: Debian GNU/Linux 8 (jessie)
* Compiler version: GCC 5.4.0
* CUDA version: 9.0
* cuDNN version: 7
* NVIDIA driver version: 381.130
* GPU models (for all devices if they are not all the same): 1 Titan Xp
* `PYTHONPATH` environment variable: No
* `python --version` output: Python 2.7.15
* Anything else that seems relevant: No
